File: ios/LineLoginModule.swift
import Foundation
import LineSDK

@objc(LineLogin) public class LineLogin: NSObject {

  private let loginLock = NSLock()
  private var isLoginInProgress = false

  @objc public static func application(
    _ application: UIApplication,
    open url: URL,
    options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool
  {
    return LoginManager.shared.application(application, open: url, options: options)
  }

  @objc public static func application(
    _ application: UIApplication,
    continue userActivity: NSUserActivity,
    restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void) -> Bool
  {
    return LoginManager.shared.application(application, open: userActivity.webpageURL)
  }

  @objc func setup(_ arguments: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock,
                   rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let channelID = arguments["channelId"] as? String, !channelID.isEmpty else {
      reject("INVALID_ARGUMENTS", "channelId must be a non-empty string", nil)
      return
    }

    guard !LoginManager.shared.isSetupFinished else {
      resolve(nil)
      return
    }

    let universalLinkURL: URL? = (arguments["universalLinkUrl"] as? String).flatMap { URL(string: $0) }
    LoginManager.shared.setup(channelID: channelID, universalLinkURL: universalLinkURL)
    resolve(nil)
  }

  @objc func login(_ arguments: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock,
                   rejecter reject: @escaping RCTPromiseRejectBlock) {
    loginLock.lock()
    guard !isLoginInProgress else {
      loginLock.unlock()
      reject("LOGIN_IN_PROGRESS", "A login is already in progress", nil)
      return
    }
    isLoginInProgress = true
    loginLock.unlock()

    let scopes = (arguments["scopes"] as? [String])?.map { LoginPermission(rawValue: $0) } ?? [.profile]
    let onlyWebLogin = (arguments["onlyWebLogin"] as? Bool) ?? false
    var parameters = LoginManager.Parameters()

    if onlyWebLogin { parameters.onlyWebLogin = onlyWebLogin }

    if let botPromptRaw = arguments["botPrompt"] as? String {
      switch botPromptRaw {
      case "aggressive": parameters.botPromptStyle = LoginManager.BotPrompt(rawValue: "aggressive")
      case "normal":     parameters.botPromptStyle = LoginManager.BotPrompt(rawValue: "normal")
      default:
        loginLock.lock()
        isLoginInProgress = false
        loginLock.unlock()
        reject("INVALID_ARGUMENT", "Invalid botPrompt '\(botPromptRaw)'. Expected: 'normal' or 'aggressive'", nil)
        return
      }
    }

    DispatchQueue.main.async {
      LoginManager.shared.login(
        permissions: Set(scopes),
        in: nil,
        parameters: parameters) { [weak self] result in
          guard let self else { return }
          self.loginLock.lock()
          self.isLoginInProgress = false
          self.loginLock.unlock()
          switch result {
          case .success(let value): resolve(self.parseLoginResult(value))
          case .failure(let error): error.rejecter(reject)
          }
        }
    }
  }

  @objc func logout(_ resolve: @escaping RCTPromiseResolveBlock,
                    rejecter reject: @escaping RCTPromiseRejectBlock) {
    LoginManager.shared.logout { result in
      switch result {
      case .success: resolve(nil)
      case .failure(let error): error.rejecter(reject)
      }
    }
  }

  @objc func getCurrentAccessToken(_ resolve: @escaping RCTPromiseResolveBlock,
                                   rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let token = AccessTokenStore.shared.current else {
      reject("ACCESS_TOKEN_NOT_AVAILABLE", "No access token is available", nil)
      return
    }
    resolve(self.parseAccessToken(token))
  }

  @objc func getFriendshipStatus(_ resolve: @escaping RCTPromiseResolveBlock,
                                 rejecter reject: @escaping RCTPromiseRejectBlock) {
    API.getBotFriendshipStatus { result in
      switch result {
      case .success(let status): resolve(self.parseFriendshipStatus(status))
      case .failure(let error): error.rejecter(reject)
      }
    }
  }

  @objc func getProfile(_ resolve: @escaping RCTPromiseResolveBlock,
                        rejecter reject: @escaping RCTPromiseRejectBlock) {
    API.getProfile { result in
      switch result {
      case .success(let profile): resolve(self.parseProfile(profile))
      case .failure(let error): error.rejecter(reject)
      }
    }
  }

  @objc func refreshAccessToken(_ resolve: @escaping RCTPromiseResolveBlock,
                                rejecter reject: @escaping RCTPromiseRejectBlock) {
    API.Auth.refreshAccessToken { result in
      switch result {
      case .success(let token): resolve(self.parseAccessToken(token))
      case .failure(let error): error.rejecter(reject)
      }
    }
  }

  @objc func verifyAccessToken(_ resolve: @escaping RCTPromiseResolveBlock,
                               rejecter reject: @escaping RCTPromiseRejectBlock) {
    API.Auth.verifyAccessToken { result in
      switch result {
      case .success(let token): resolve(self.parseVerifyAccessToken(token))
      case .failure(let error): error.rejecter(reject)
      }
    }
  }

  private func parseFriendshipStatus(_ status: GetBotFriendshipStatusRequest.Response) -> NSDictionary {
    return [
      "friendFlag": status.friendFlag
    ]
  }

  private func parseAccessToken(_ accessToken: AccessToken) -> NSDictionary {
    var result: [String: Any] = [
      "accessToken": accessToken.value,
      "expiresIn":   Int(accessToken.expiresAt.timeIntervalSinceNow),
    ]
    if let idToken = accessToken.IDTokenRaw {
      result["idToken"] = idToken
    }
    return NSDictionary(dictionary: result)
  }

  private func parseProfile(_ profile: UserProfile) -> NSDictionary {
    var result: [String: Any] = [
      "displayName": profile.displayName,
      "userId":      profile.userID,
    ]
    if let url = profile.pictureURL {
      result["pictureUrl"] = url.absoluteString
    }
    if let message = profile.statusMessage {
      result["statusMessage"] = message
    }
    return NSDictionary(dictionary: result)
  }

  private func parseLoginResult(_ loginResult: LoginResult) -> NSDictionary {
    var result: [String: Any] = [
      "accessToken": self.parseAccessToken(loginResult.accessToken),
      "scope":       loginResult.permissions.map { $0.rawValue }.joined(separator: " "),
    ]
    if let changed = loginResult.friendshipStatusChanged {
      result["friendshipStatusChanged"] = changed
    }
    if let nonce = loginResult.IDTokenNonce {
      result["idTokenNonce"] = nonce
    }
    if let profile = loginResult.userProfile {
      result["userProfile"] = self.parseProfile(profile)
    }
    return NSDictionary(dictionary: result)
  }

  private func parseVerifyAccessToken(_ accessToken: AccessTokenVerifyResult) -> NSDictionary {
    return [
      "clientId":  accessToken.channelID,
      "expiresIn": accessToken.expiresIn,
      "scope":     accessToken.permissions.map { $0.rawValue }.joined(separator: " ")
    ]
  }
}

extension LineSDKError {
  func rejecter(_ reject: @escaping RCTPromiseRejectBlock) {
    let code = String(errorCode)
    let message = errorDescription ?? "Unknown error"
    reject(code, message, self)
  }
}

File: android/src/main/java/com/xmartlabs/line/LineLoginModule.kt
package com.xmartlabs.line

import android.app.Activity
import android.content.Intent
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

import com.facebook.react.bridge.*
import com.facebook.react.module.annotations.ReactModule

import com.linecorp.linesdk.*
import com.linecorp.linesdk.api.LineApiClient
import com.linecorp.linesdk.api.LineApiClientBuilder
import com.linecorp.linesdk.auth.LineAuthenticationConfig
import com.linecorp.linesdk.auth.LineAuthenticationParams
import com.linecorp.linesdk.auth.LineLoginApi
import com.linecorp.linesdk.auth.LineLoginResult

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

@ReactModule(name = LineLoginModule.NAME)
class LineLoginModule(reactContext: ReactApplicationContext) :
    NativeLineLoginSpec(reactContext) {

    companion object {
        const val BOT_PROMPT = "botPrompt"
        const val CHANNEL_ID = "channelId"
        const val ONLY_WEB_LOGIN = "onlyWebLogin"
        const val SCOPES = "scopes"

        const val NAME = NativeLineLoginSpec.NAME
        private const val LOGIN_REQUEST_CODE = 1423
        private val DEFAULT_SCOPES = listOf("profile")
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val pendingLogin = AtomicReference<Promise?>(null)

    @Volatile private var channelId: String? = null
    @Volatile private var lineApiClient: LineApiClient? = null

    private val activityListenerRegistered = AtomicBoolean(false)

    override fun invalidate() {
        scope.cancel()
        pendingLogin.getAndSet(null)
            ?.reject("MODULE_INVALIDATED", "Module was destroyed during login", null)
        if (activityListenerRegistered.getAndSet(false)) {
            reactApplicationContext.removeActivityEventListener(loginResultListener)
        }
        super.invalidate()
    }

    override fun setup(args: ReadableMap, promise: Promise) {
        val id = args.getString(CHANNEL_ID)?.takeIf { it.isNotBlank() }
            ?: return promise.reject("SETUP_FAILED", "channelId must be a non-empty string", null)

        channelId = id

        lineApiClient = LineApiClientBuilder(reactApplicationContext, id).build()

        if (activityListenerRegistered.compareAndSet(false, true)) {
            reactApplicationContext.addActivityEventListener(loginResultListener)
        }
        promise.resolve(null)
    }

    override fun login(args: ReadableMap, promise: Promise) {
        requireClient(promise) ?: return

        val id = channelId ?: return promise.reject("NOT_SETUP", "Call setup() first", null)

        val scopes = args.getArray(SCOPES)
            ?.toArrayList()?.filterIsInstance<String>()?.takeIf { it.isNotEmpty() }
            ?: DEFAULT_SCOPES

        val onlyWebLogin = args.hasKey(ONLY_WEB_LOGIN) && args.getBoolean(ONLY_WEB_LOGIN)

        val botPromptRaw = args.getString(BOT_PROMPT) ?: "normal"
        val botPrompt = LineAuthenticationParams.BotPrompt.entries
            .find { it.name.equals(botPromptRaw, ignoreCase = true) }
            ?: return promise.reject(
                "INVALID_ARGUMENT",
                "Invalid botPrompt '$botPromptRaw'. Expected: ${
                    LineAuthenticationParams.BotPrompt.entries.joinToString { it.name.lowercase() }
                }",
                null,
            )

        val activity = currentActivity
            ?: return promise.reject("NO_ACTIVITY", "Activity is not available", null)

        if (!pendingLogin.compareAndSet(null, promise)) {
            return promise.reject("LOGIN_IN_PROGRESS", "A login is already in progress", null)
        }

        val authParams = LineAuthenticationParams.Builder()
            .scopes(Scope.convertToScopeList(scopes))
            .botPrompt(botPrompt)
            .build()

        val config = if (onlyWebLogin) {
            LineAuthenticationConfig.Builder(id).disableLineAppAuthentication().build()
        } else null

        val intent = if (config != null) {
            LineLoginApi.getLoginIntent(activity, config, authParams)
        } else {
            LineLoginApi.getLoginIntent(activity, id, authParams)
        }

        activity.startActivityForResult(intent, LOGIN_REQUEST_CODE)
    }

    private val loginResultListener = object : ActivityEventListener {
        override fun onNewIntent(intent: Intent) = Unit

        override fun onActivityResult(
            activity: Activity,
            requestCode: Int,
            resultCode: Int,
            data: Intent?
        ) {
            if (requestCode != LOGIN_REQUEST_CODE) return
            val promise = pendingLogin.getAndSet(null) ?: return

            if (resultCode != Activity.RESULT_OK || data == null) {
                val (code, message) = if (resultCode == Activity.RESULT_CANCELED) {
                    "LOGIN_CANCELLED" to "Login was cancelled by the user"
                } else {
                    "LOGIN_FAILED" to "Login failed with unexpected result code $resultCode"
                }
                return promise.reject(code, message, null)
            }

            val result = LineLoginApi.getLoginResultFromIntent(data)
            if (result.responseCode == LineApiResponseCode.SUCCESS) {
                val cred = result.lineCredential
                val prof = result.lineProfile
                if (cred != null && prof != null) {
                    promise.resolve(buildLoginResult(result, cred, prof))
                } else {
                    promise.reject(
                        "PARSE_ERROR",
                        "Credential or profile missing from login result",
                        null
                    )
                }
            } else {
                promise.reject(
                    result.responseCode.name,
                    result.errorData.message ?: result.responseCode.name,
                    null
                )
            }
        }
    }

    override fun logout(promise: Promise) {
        val client = requireClient(promise) ?: return
        scope.launch {
            try {
                val r = client.logout()
                if (r.isSuccess) promise.resolve(null)
                else promise.reject(
                    r.responseCode.name,
                    r.errorData.message ?: r.responseCode.name,
                    null
                )
            } catch (e: Exception) {
                promise.reject("API_ERROR", e.message ?: "Unexpected error during logout", e)
            }
        }
    }

    override fun getProfile(promise: Promise) = apiCall(
        promise,
        call = LineApiClient::getProfile,
        build = ::buildProfile,
    )

    override fun getCurrentAccessToken(promise: Promise) = apiCall(
        promise,
        call = LineApiClient::getCurrentAccessToken,
        build = { buildAccessToken(it, idToken = null) },
    )

    override fun getFriendshipStatus(promise: Promise) = apiCall(
        promise,
        call = LineApiClient::getFriendshipStatus,
        build = ::buildFriendshipStatus,
    )

    override fun refreshAccessToken(promise: Promise) = apiCall(
        promise,
        call = LineApiClient::refreshAccessToken,
        build = { buildAccessToken(it, idToken = null) },
    )

    override fun verifyAccessToken(promise: Promise) = apiCall(
        promise,
        call = LineApiClient::verifyToken,
        build = ::buildVerifyResult,
    )

    /**
     * Returns the initialized [LineApiClient] or rejects [promise] with NOT_SETUP
     * and returns null, allowing callers to use the `?: return` pattern.
     */
    private fun requireClient(promise: Promise): LineApiClient? =
        lineApiClient ?: run {
            promise.reject("NOT_SETUP", "Call setup() before using the LINE SDK", null)
            null
        }

    /**
     * Runs a blocking LINE API [call] on the IO dispatcher.
     * Resolves [promise] via [build] on success, or rejects it with the LINE error.
     */
    private fun <T> apiCall(
        promise: Promise,
        call: (LineApiClient) -> LineApiResponse<T>,
        build: (T) -> WritableMap,
    ) {
        val client = requireClient(promise) ?: return
        scope.launch {
            try {
                val r = call(client)
                if (r.isSuccess) promise.resolve(build(r.responseData))
                else promise.reject(
                    r.responseCode.name,
                    r.errorData.message ?: r.responseCode.name,
                    null
                )
            } catch (e: Exception) {
                promise.reject("API_ERROR", e.message ?: "Unexpected error", e)
            }
        }
    }

    /**
     * [expiresIn] is seconds-until-expiry (OAuth standard `expires_in`),
     * derived from the SDK's millisecond value.
     */
    private fun buildAccessToken(token: LineAccessToken, idToken: LineIdToken?): WritableMap =
        Arguments.makeNativeMap(
            mapOf(
                "accessToken" to token.tokenString,
                "expiresIn"   to token.expiresInMillis / 1000L,
                "idToken"     to idToken?.rawString,
            )
        )

    private fun buildProfile(profile: LineProfile): WritableMap =
        Arguments.makeNativeMap(
            mapOf(
                "displayName"   to profile.displayName,
                "pictureUrl"    to profile.pictureUrl?.toString(),
                "statusMessage" to profile.statusMessage,
                "userId"        to profile.userId,
            )
        )

    private fun buildFriendshipStatus(status: LineFriendshipStatus): WritableMap =
        Arguments.makeNativeMap(
            mapOf(
                "friendFlag" to status.isFriend
            )
        )

    private fun buildLoginResult(result: LineLoginResult, credential: LineCredential, profile: LineProfile): WritableMap =
        Arguments.makeNativeMap(
            mapOf(
                "accessToken"             to buildAccessToken(credential.accessToken, result.lineIdToken),
                "friendshipStatusChanged" to result.friendshipStatusChanged,
                "idTokenNonce"            to result.lineIdToken?.let { result.nonce },
                "scope"                   to Scope.join(credential.scopes),
                "userProfile"             to buildProfile(profile),
            )
        )

    private fun buildVerifyResult(credential: LineCredential): WritableMap =
        Arguments.makeNativeMap(
            mapOf(
                "clientId"  to channelId,
                "expiresIn" to credential.accessToken.expiresInMillis / 1000L,
                "scope"     to Scope.join(credential.scopes),
            )
        )
}
